extensions/pdh/PerformanceDataMonitor.cpp

```diff
@@ -0,0 +1,283 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PerformanceDataMonitor.h"
+#include "PDHCounters.h"
+#include "MemoryConsumptionCounter.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+core::Relationship PerformanceDataMonitor::Success("success", "All files are routed to success");
+
+core::Property PerformanceDataMonitor::PredefinedGroups(
+    core::PropertyBuilder::createProperty("Predefined Groups")->
+    withDescription("Comma separated list to which predefined groups should be included")->
+    withDefaultValue("")->build());
+
+core::Property PerformanceDataMonitor::CustomPDHCounters(
+    core::PropertyBuilder::createProperty("Custom PDH Counters")->
+    withDescription("Comma separated list of PDHCounters to collect from")->
+    withDefaultValue("")->build());
+
+core::Property PerformanceDataMonitor::OutputFormatProperty(
+    core::PropertyBuilder::createProperty("Output format")->
+    withDescription("Format of the created flowfiles")->
+    withAllowableValue<std::string>(JSON_FORMAT_STR)->
+    withAllowableValue(OPEN_TELEMETRY_FORMAT_STR)->
+    withDefaultValue(JSON_FORMAT_STR)->build());
+
+PerformanceDataMonitor::~PerformanceDataMonitor() {
+  if (pdh_query_ != nullptr)
+    PdhCloseQuery(pdh_query_);
+  for (PerformanceDataCounter* resource_consumption_counter : resource_consumption_counters_) {
+    delete resource_consumption_counter;
+  }
+}
+
+void PerformanceDataMonitor::onSchedule(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSessionFactory>& sessionFactory) {
+  setupMembersFromProperties(context);
+
+  std::vector<PerformanceDataCounter*> valid_counters;
+
+  for (PerformanceDataCounter* counter : resource_consumption_counters_) {
+    PDHCounterBase* pdh_counter = dynamic_cast<PDHCounterBase*> (counter);
+    if (pdh_counter != nullptr) {
+      if (pdh_query_ == nullptr)
+        PdhOpenQuery(NULL, NULL, &pdh_query_);
+      PDH_STATUS add_to_query_result = pdh_counter->addToQuery(pdh_query_);
+      if (add_to_query_result != ERROR_SUCCESS) {
+        logger_->log_error(("Error adding " + pdh_counter->getName() + " to query: " + std::to_string(add_to_query_result)).c_str());
+        delete counter;
+      } else {
+        valid_counters.push_back(counter);
+      }
+    } else {
+      valid_counters.push_back(counter);
+    }
+  }
```

Comment:
it seems to me that this loop is the reason we use raw pointers, `std::vector<std::unique_ptr<PerformanceDataCounter>>` could be used with `remove_if-erase` to get rid of the raw pointers (if indeed this is the only reason)

Reply:
changed the raw ptrs to be unique_ptr-s in  [d5959c3](https://github.com/martinzink/nifi-minifi-cpp/commit/d5959c3b1b5d0f377e7b7f2c13cc22a484b2e4a6)